### PR TITLE
feat: build nix image

### PR DESCRIPTION
**Pull request by lostbean**
## Description:
Add ability to build Nix images inside APIC.

## Is this change user facing?
YES

### CODE REVIEW

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/kubernetes_kurtosis_backend.go [hunk 1]
```diff
@@ -487,6 +489,11 @@ func (backend *KubernetesKurtosisBackend) BuildImage(ctx context.Context, imageN
 	return "", stacktrace.NewError("Building images isn't yet implemented in Kubernetes.")
 }
 
+func (backend *KubernetesKurtosisBackend) NixBuild(ctx context.Context, nixBuildSpec *nix_build_spec.NixBuildSpec) (string, error) {
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/kubernetes_kurtosis_backend.go:492]: Could you add this diff on this doc https://www.notion.so/kurtosistech/Docker-Kubernetes-Inconsistencies-Discrepancies-ff62b3970c5e44fab4984635459fc825 ?

File: container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go [hunk 2]
```diff
@@ -1313,6 +1322,108 @@ func (manager *DockerManager) FetchImage(ctx context.Context, image string, regi
...
+	var found bool = false
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, stacktrace.Propagate(err, "Fail to read the image files")
+		}
+		if header.Name == "manifest.json" {
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go]: hardcoded value

File: container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go [hunk 3]
```diff
@@ -1313,6 +1322,108 @@ func (manager *DockerManager) FetchImage(ctx context.Context, image string, regi
...
+	imageFileRaw, err := cmd.Output()
+	if err != nil {
+		errMsg := errBuffer.String()
+		logrus.WithError(err).Error(errMsg)
+		return "", stacktrace.Propagate(err, "Failed to build nix image with Nix.")
+	}
+	imageFile := strings.TrimSpace(string(imageFileRaw))
+	logrus.Debugf("Nix flake image on attribute %s, result on image file %s", flakeReference, imageFile)
+
+	imageName, err := GetRepoTags(imageFile)
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go]: nit: the name var could be renamed `imageTags`

File: container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go [hunk 4]
```diff
@@ -1313,6 +1322,108 @@ func (manager *DockerManager) FetchImage(ctx context.Context, image string, regi
...
+		return "", stacktrace.Propagate(err, "Failed to open generated Nix image on %s", imageFile)
+	}
+
+	_, err = manager.dockerClient.ImageLoad(ctx, image, false)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "Failed to load Nix image %s in docker", imageFile)
+	}
+	logrus.Debugf("Nix generated image file %s is loaded into docker", imageFile)
+
+	return imageName[0], nil
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go]: is the image name always the first tag?

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/kubernetes_kurtosis_backend.go [hunk 5]
```diff
@@ -487,6 +489,11 @@ func (backend *KubernetesKurtosisBackend) BuildImage(ctx context.Context, imageN
...
+	// TODO IMPLEMENT
+	return "", stacktrace.NewError("Loading images isn't yet implemented in Kubernetes.")
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/kubernetes_kurtosis_backend.go]:
```suggestion
	return "", stacktrace.NewError("Nix builds isn't yet implemented in Kubernetes.")
```

File: container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go (hunk 4 above)
Comment (leoporoli) [reply container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go]: is it possible to add an extra guardrail for this?

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/nix_build_spec.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+const (
+	NixBuildSpecTypeName = "NixBuildSpec"
+
+	FlakeLocationDir = "flake_location_dir"
+	FlakeOutputAttr  = "flake_output"
+	NixContextAttr   = "build_context_dir"
+	NixImageName     = "image_name"
+
+	// Currently only supports container nixs named Dockerfile
```
Comment (leoporoli) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/nix_build_spec.go]: is this comment ok?

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/nix_build_spec.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+	if interpretationErr != nil {
+		return "", "", interpretationErr
+	}
+
+	contextDirOnDisk, interpretationErr := packageContentProvider.GetOnDiskAbsolutePath(contextDirAbsoluteLocator)
+	if interpretationErr != nil {
+		return "", "", interpretationErr
+	}
+	// Assume, that flake nix sits at the same level as context directory to get context dir path on disk
+	FlakeDirOnDisk := filepath.Dir(flakeNixPathOnDisk)
```
Comment (leoporoli) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/nix_build_spec.go]: name var starting with uppercase

File: docs/docs/api-reference/starlark-reference/nix-support.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+You can provide Kurtosis just with source code and a Nix definition on how to build an image. Kurtosis will take care of building and deploying the image directly into the enclave without the need to upload or register the image beforehand.
+
+For that, we use Nix flakes, which is a way to package build definitions and dependencies in a reproducible manner. Using Nix flakes, you can define your system configurations and dependencies in a single file (`flake.nix`), making it easier to manage and share.
+
+Here's a basic explanation of how you can generate Docker images from services using Nix flakes:
+
+1. **Install Nix**: Installing Nix isn't strictly necessary with Kurtosis, but it's recommended if you are creating or developing the package. You can install it by following the instructions on the Nix website: [https://nixos.org/download.html](https://nixos.org/download.html)
+
+2. **Create a Nix Flake**: Create a directory for your project and initialize a Nix flake. You can do this by running:
```
Comment (leoporoli) [inline docs/docs/api-reference/starlark-reference/nix-support.md]: Should we also clarify that this folder should contain the service's source code here? or is it very clear?

File: docs/docs/api-reference/starlark-reference/nix-support.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        name = "nix-example",
+        config = ServiceConfig(
+            image = NixBuildSpec(image_name = "myservice", flake_location_dir = ".", build_context_dir = "./"),
+        ),
+    )
+   ```
+
+5. **Build and Deploy with Kurtosis**: From your package folder, simply run `kurtosis run .` to get your cluster up and running.
+
+This is just a basic example. Depending on your specific use case and requirements, you may need to adjust the configuration and dependencies in your `flake.nix` file accordingly. Additionally, you can add more services, configure networking, volumes, environment variables, etc., based on your needs.
```
Comment (leoporoli) [inline docs/docs/api-reference/starlark-reference/nix-support.md:52]: Nice!

File: docs/docs/api-reference/starlark-reference/service-config.md [hunk 10]
```diff
@@ -52,6 +52,26 @@ config = ServiceConfig(
...
+    
+    OR
+
+    image = NixBuildSpec(
+        #  The name of the image that needs to be pulled qualified with the registry
+        # MANDATORY
+        image_name = "hello-world-server",
+
+        # Locator to build context within the Kurtosis package
+        # As of now, Kurtosis expects a Dockerfile at the root of the build context
```
Comment (leoporoli) [inline docs/docs/api-reference/starlark-reference/service-config.md]: is Dockerfile ok here?

**Review by leoporoli — APPROVED** [review]

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/kubernetes_kurtosis_backend.go (hunk 1 above)
Comment (lostbean) [reply container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/kubernetes_kurtosis_backend.go:492]: Added!

File: container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go (hunk 4 above)
Comment (lostbean) [reply container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go]: yeah, hard to guarantee that but we definitely need to add some checks

File: docs/docs/api-reference/starlark-reference/nix-support.md (hunk 8 above)
Comment (lostbean) [reply docs/docs/api-reference/starlark-reference/nix-support.md]: yeah, not very clear. It is implicitly assuming that you are starting a project and that later you will put the source code there. Maybe a more realistic version would be:
> Go to your project root directory and initialize a Nix flake

File: docs/docs/api-reference/starlark-reference/service-config.md (hunk 10 above)
Comment (lostbean) [reply docs/docs/api-reference/starlark-reference/service-config.md]: nope!

**Pull request merged**